### PR TITLE
Feature/utility contract

### PR DESCRIPTION
**Pull request by myronrotter**
Add Utility interface `IUtility.sol`. `Utility.sol` implements `IUtility.sol`.
Tracking of enery production and consumption is implemented. Specific tracking events are emitted.
Energy production and consumption is aggregated and stored on the chain.
Settlement (netting) is not implemented yet.

Furthermore, `Utility.sol` inherits from (basic contracts). `Mortal.sol` and `Owned.sol`.
See and execute `utility.test.js` for testing `Ulitity.sol`.

### CODE REVIEW

File: contracts/IUtility.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  event NonRenewableEnergyChanged(address indexed household, int256 energy);
+
+  function addHousehold(address _household) external returns (bool);
+
+  function updateRenewableEnergy(address _household, int256 _producedEnergy, int256 _consumedEnergy) external returns (bool);
+
+  function updateNonRenewableEnergy(address _household, int256 _producedEnergy, int256 _consumedEnergy) external returns (bool);
+
+  function settle() external returns (bool);
```
Comment (dakingha69) [inline contracts/IUtility.sol]: Any specific reason why you chose `settle` instead of `netting`?

File: contracts/Owned.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @title Owned
+ * @notice Simple owner contract.
+ * @dev Is inherited by derived contracts.
+ */
+contract Owned {
+  address payable owner;
+
+  constructor (/*address _owner*/) public {
```
Comment (dakingha69) [inline contracts/Owned.sol]: If we are not using this constructor argument then you can remove it IMO.

File: contracts/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+- `Owned.sol` and `Mortal.sol`: defines ownership of the utility contract and has the authority to permanently deactivate it.
+- `Utility.sol`: the core logic of our utility that households want to talk to. Implements `IUtility.sol`.
+
+## Common functions
+
+The following is an outline of how we expect the flow of function calls looks like:
+
+1. households periodically provide information about energy production and consumption by using `updateRenewableEnergy` and/or `updateNonRenewableEnergy`.
+2. (WIP) authority nodes periodically invoke regulation of the produce/consume imbalance such that most households are *happy* (`settle`).
+3. (WIP) households that produced energy can receive their reward by checking if they have successfully distributed energy. The reward comes in the form of an event, e.g, `TransferEnergy`. An alternative approach to this is emitting `TransferEnergy` immediately inside `settle` for each transfer.
```
Comment (dakingha69) [inline contracts/README.md]: What do you mean by `reward`?

File: contracts/Utility.sol [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+/**
+ * @title Utility
+ * @notice Tracks production/consumption of energy of all households. Settles energy requests by distributing existing energy as fair as possible (netting).
+ * @dev Implements interface IUtility.
+ */
+contract Utility is IUtility, Mortal {
+  /*
+   * electrical work/energy W, unit 1 kWh = 1000 Wh = 1000 W * 3600 s = 3,6 * 10^6 Ws
+   * (int256) 1 means 1 Ws?
```
Comment (dakingha69) [inline contracts/Utility.sol]: Yes I think it makes sense to use Ws in the contract.

File: contracts/Utility.sol [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  // mapping of all households
+  mapping(address => Household) households;
+
+  modifier onlyHousehold(address _household) {
+    require(msg.sender == _household, "No permission to access. Only household can access itself.");
+    _;
+  }
+
+  // modifier such that only some authority can access functions
+  modifier onlyEnergyAuthority() {
```
Comment (dakingha69) [inline contracts/Utility.sol]: What is the difference between this modifier and `onlyHousehold`?

File: contracts/Utility.sol [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    _;
+  }
+
+  modifier householdExists(address _household) {
+    require(households[_household].initialized, "Household does not exist.");
+    _;
+  }
+
+  /* solium-disable-next-line */
+  constructor() public Owned() {
```
Comment (dakingha69) [inline contracts/Utility.sol]: Is this necessary? Can't you just remove the constructor if it is empty anyways?

File: test/utility.test.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  });
+
+  describe("Households", () => {
+    context("with a new household", async () => {
+      beforeEach(async () => {
+        ({ logs: this.logs } = await this.instance.addHousehold(household));
+      });
+
+      it("should create a new household", async () => {
+        let hh = await this.instance.getHousehold(household);
```
Comment (dakingha69) [inline test/utility.test.js]: Should be `const`

File: contracts/IUtility.sol (hunk 1 above)
Comment (minhi) [reply contracts/IUtility.sol]: To be honest, personal preference. When Jacob spoke about settle/settlement it sticked that way. The term 'settle' sounds like a more general term than 'netting' though.

File: contracts/Utility.sol (hunk 5 above)
Comment (minhi) [reply contracts/Utility.sol]: None. I think it was some leftovers from the first few commits. Will remove.

File: contracts/README.md (hunk 3 above)
Comment (minhi) [reply contracts/README.md]: Reward in the sense of producers getting rewarded from distributing their energy per the netting algorithm. The event `TransferEnergy` should be like a proof recorded on the blockchain that producer A transferred X amount of energy to consumer B. What I'm imagining coming next is some authority looking looking through those events and enforces that B pays A accordingly.

**Pull request merged**